appserver/web/war-util/src/test/java/org/glassfish/web/loader/WebappClassLoaderTest.java

```diff
@@ -0,0 +1,183 @@
+    private File junitJarFile;
+
+    @Before
+    public void setup() throws URISyntaxException {
+        // Run 3 methods at the same time, and make the pool large enough to increase
+        // the chance of a race condition
+        executor = Executors.newFixedThreadPool(60);
+
+        // Require a minimum number of executions before completing
+        latch = new CountDownLatch(EXECUTION_COUNT);
+
+        // Fetch any JAR to use for classloading
+        junitJarFile = new File(Test.class.getProtectionDomain().getCodeSource().getLocation().toURI());
+    }
+
+    @After
+    public void shutdown() throws InterruptedException {
+        if (executor != null) {
+            executor.shutdownNow();
+            assertTrue("Executor could not shutdown. This could mean there is a deadlock.",
+                    executor.awaitTermination(10, TimeUnit.SECONDS));
+        }
+    }
+
+    @Test
+    public void check_findResourceInternalFromJars_thread_safety() throws Exception {
+        final WebappClassLoader webappClassLoader = new WebappClassLoader(getClass().getClassLoader(), null);
+        webappClassLoader.start();
+        webappClassLoader.setResources(new FileDirContext());
+
+        CompletableFuture<Void> result = new CompletableFuture<>();
+
+        // Create the tasks to run
+        Runnable lookupTask = waitAndDo(result, () -> lookup(webappClassLoader));
+        Runnable addTask = waitAndDo(result, () -> add(webappClassLoader));
+        Runnable closeTask = waitAndDo(result, () -> webappClassLoader.closeJARs(true));
+
+        try {
+            // Run the methods at the same time
+            for (int i = 0; i < EXECUTION_COUNT; i++) {
+                executor.execute(addTask);
+                executor.execute(lookupTask);
+                executor.execute(closeTask);
+            }
+
+            // Wait for tasks to execute
+            assertTrue("The tasks didn't finish in the allowed time.",
+                    latch.await(20, TimeUnit.SECONDS));
+
+            // Check to see if any tasks completed exceptionally
+            result.getNow(null);
+        } finally {
+            webappClassLoader.close();
+        }
+    }
+
+    private void add(WebappClassLoader webappClassLoader) throws IOException {
+        List<JarFile> jarFiles = findJarFiles();
+
+        for (JarFile j : jarFiles) {
+            try {
+                webappClassLoader.addJar(junitJarFile.getName(), j, junitJarFile);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private void lookup(WebappClassLoader webappClassLoader) throws Exception {
+        for (JarFile jarFile : findJarFiles()) {
+            for (JarEntry entry : Collections.list(jarFile.entries())) {
+                webappClassLoader.findResource(entry.getName());
+                // System.out.println("Looked up " + resourceEntry);
+                Thread.sleep(0, 100);
```

Comment:
This sleep means that only a 'few' lookups are performed (in my test, around 3000 in the 130 ms the test is running).

By replacing it with `Thread.yield()`, the number of lookup actions is increased to around 21000 which mean the chance of having a concurrency issue (which we like to detect with this test) is increased significantly.

Reply:
`Thread.yield()` causes a huge amount of false positives on the test. I imagine this is because it reduces the total time spent on this task